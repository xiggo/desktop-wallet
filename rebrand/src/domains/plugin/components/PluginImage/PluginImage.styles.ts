import tw from "twin.macro";

import { Size } from "@/types";

const baseStyle = tw`flex flex-col space-y-3 items-center justify-center transition-all duration-200 overflow-hidden`;

const getSize = (size?: Size) => {
	const sizes = {
		"2xs": () => tw`flex-shrink-0 h-8 w-8 rounded`,
		"3xs": () => tw`flex-shrink-0 h-6 w-6 rounded`,
		default: () => tw`flex-shrink-0 h-25 w-25 rounded-xl`,
		lg: () => tw`flex-shrink-0 h-32 w-32 rounded-xl`,
		sm: () => tw`flex-shrink-0 h-15 w-15 rounded-lg`,
		xl: () => tw`flex-shrink-0 h-44 w-44 rounded-2.5xl`,
		xs: () => tw`flex-shrink-0 h-11 w-11 rounded-lg`,
	};

	// eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
	return (sizes[size as keyof typeof sizes] || sizes.default)();
};

const getVariant = (variant?: string) => {
	if (variant === "progress") {
		return tw`border border-theme-secondary-300 dark:border-theme-secondary-800 bg-theme-success-100 dark:bg-theme-success-900`;
	}
};

export const getStyles = ({ size, variant }: { size?: Size; variant?: string }) => [
	baseStyle,
	getSize(size),
	getVariant(variant),
];
