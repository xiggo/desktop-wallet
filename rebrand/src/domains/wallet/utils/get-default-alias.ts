import { Networks } from "@payvo/sdk";
import { Contracts } from "@payvo/sdk-profiles";

interface GetDefaultAliasInput {
	profile: Contracts.IProfile;
	network: Networks.Network;
}

export const getDefaultAlias = ({ profile, network }: GetDefaultAliasInput): string => {
	const makeAlias = (count: number) => `${network.displayName()} #${count}`;

	const sameCoinWallets = profile.wallets().findByCoinWithNetwork(network.coin(), network.id());

	let counter = sameCoinWallets.length;

	if (counter === 0) {
		counter = 1;
	}

	while (profile.wallets().findByAlias(makeAlias(counter))) {
		counter++;
	}

	return makeAlias(counter);
};
