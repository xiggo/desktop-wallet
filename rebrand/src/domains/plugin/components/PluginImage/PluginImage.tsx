import cn from "classnames";
import React, { useMemo, useState } from "react";
import { useTranslation } from "react-i18next";
import { styled } from "twin.macro";

import { getStyles } from "./PluginImage.styles";
import { CircularProgressBar } from "@/app/components/CircularProgressBar";
import { Image } from "@/app/components/Image";
import { useTheme } from "@/app/hooks";
import { Size } from "@/types";

export enum PluginImageProgressSize {
	Small = 20,
	Base = 40,
}

interface Properties {
	size?: Size;
	logoURL?: string;
	isUpdating?: boolean;
	updatingProgress?: number;
	progressSize?: PluginImageProgressSize;
	className?: string;
	showUpdatingLabel?: boolean;
}

const PluginImageWrapper = styled.div<{ size?: Size; variant?: string }>(getStyles);

export const PluginImage = ({
	size,
	logoURL,
	isUpdating,
	updatingProgress = 0,
	progressSize = PluginImageProgressSize.Base,
	className,
	showUpdatingLabel,
}: Properties) => {
	const { isDarkMode } = useTheme();

	const { t } = useTranslation();

	const colors = useMemo(
		() => ({
			progressColor: "var(--theme-color-success-600)",
			strokeColor: isDarkMode ? "var(--theme-color-success-800)" : "var(--theme-color-success-200)",
		}),
		[isDarkMode],
	);

	const [hasError, setHasError] = useState(false);

	const renderContent = () => {
		if (isUpdating) {
			return (
				<>
					<CircularProgressBar
						value={+(updatingProgress * 100).toFixed(0)}
						size={progressSize}
						strokeWidth={3}
						fontSize={0.8}
						showValue={progressSize !== PluginImageProgressSize.Small}
						{...colors}
					/>
					{showUpdatingLabel && (
						<p
							data-testid="PluginImage__updating__label"
							className="text-sm font-semibold text-theme-success-600"
						>
							{t("COMMON.UPDATING")}
						</p>
					)}
				</>
			);
		}

		if (hasError || !logoURL) {
			return <Image name="PluginLogoPlaceholder" domain="plugin" />;
		}

		return (
			<img src={logoURL} alt="Logo" className="object-cover w-full h-full" onError={() => setHasError(true)} />
		);
	};

	return (
		<PluginImageWrapper
			size={size}
			variant={isUpdating ? "progress" : undefined}
			className={cn(
				"bg-theme-primary-100 text-theme-primary-600 dark:bg-theme-secondary-800 dark:text-theme-secondary-700",
				className,
			)}
			data-testid="PluginImage"
		>
			{renderContent()}
		</PluginImageWrapper>
	);
};
