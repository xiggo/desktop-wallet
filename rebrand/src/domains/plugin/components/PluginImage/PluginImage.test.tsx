import React from "react";

import { PluginImage } from "./PluginImage";
import { Size } from "@/types";
import * as utils from "@/utils/electron-utils";
import { fireEvent, render, screen } from "@/utils/testing-library";

describe("PluginImage", () => {
	it("should render image placeholder", () => {
		const { container } = render(<PluginImage />);

		expect(screen.getByTestId("PluginImage")).toHaveTextContent("plugin-logo-placeholder.svg");
		expect(container).toMatchSnapshot();
	});

	it("should render image logo", () => {
		const { container } = render(<PluginImage logoURL="https://payvo.com/logo.png" />);

		expect(screen.getByTestId("PluginImage")).not.toHaveTextContent("plugin-logo-placeholder.svg");
		expect(screen.getByTestId("PluginImage")).not.toHaveTextContent("exchange-logo-placeholder.svg");
		expect(container).toMatchSnapshot();
	});

	it.each(["dark", "light"])("should render updating element in %s mode", (theme) => {
		const utilsSpy = jest.spyOn(utils, "shouldUseDarkColors").mockImplementation(() => theme === "dark");

		const { container } = render(<PluginImage updatingProgress={25} isUpdating />);

		expect(screen.getByTestId("CircularProgressBar__percentage")).toBeInTheDocument();
		expect(container).toMatchSnapshot();

		utilsSpy.mockRestore();
	});

	it("should render updating with label", () => {
		const { container } = render(<PluginImage updatingProgress={25} isUpdating showUpdatingLabel />);

		expect(screen.getByTestId("PluginImage__updating__label")).toBeInTheDocument();
		expect(container).toMatchSnapshot();
	});

	it.each(["3xs", "2xs", "xs", "sm", "md", "lg"])("should render with size '%s'", (size) => {
		const { container } = render(<PluginImage size={size as Size} />);

		expect(screen.getByTestId("PluginImage")).toHaveTextContent("plugin-logo-placeholder.svg");
		expect(container).toMatchSnapshot();
	});

	it("should show placeholder as fallback", () => {
		const { container } = render(<PluginImage logoURL="https://payvo.com/logo.png" />);

		fireEvent.error(screen.getByRole("img"));

		expect(screen.getByTestId("PluginImage")).toHaveTextContent("plugin-logo-placeholder.svg");
		expect(container).toMatchSnapshot();
	});
});
