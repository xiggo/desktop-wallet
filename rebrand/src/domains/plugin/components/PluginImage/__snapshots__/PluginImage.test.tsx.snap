// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PluginImage should render image logo 1`] = `
<div>
  <div
    class="PluginImage__PluginImageWrapper-sc-10kztrf-0 cWrKPM bg-theme-primary-100 text-theme-primary-600 dark:bg-theme-secondary-800 dark:text-theme-secondary-700"
    data-testid="PluginImage"
  >
    <img
      alt="Logo"
      class="object-cover w-full h-full"
      src="https://payvo.com/logo.png"
    />
  </div>
</div>
`;

exports[`PluginImage should render image placeholder 1`] = `
<div>
  <div
    class="PluginImage__PluginImageWrapper-sc-10kztrf-0 cWrKPM bg-theme-primary-100 text-theme-primary-600 dark:bg-theme-secondary-800 dark:text-theme-secondary-700"
    data-testid="PluginImage"
  >
    <svg>
      plugin-logo-placeholder.svg
    </svg>
  </div>
</div>
`;

exports[`PluginImage should render updating element in dark mode 1`] = `
<div>
  <div
    class="PluginImage__PluginImageWrapper-sc-10kztrf-0 fBINnA bg-theme-primary-100 text-theme-primary-600 dark:bg-theme-secondary-800 dark:text-theme-secondary-700"
    data-testid="PluginImage"
  >
    <svg
      height="40"
      viewBox="0 0 40 40"
      width="40"
    >
      <circle
        cx="20"
        cy="20"
        fill="transparent"
        r="18.5"
        stroke="var(--theme-color-success-800)"
        stroke-width="3"
      />
      <circle
        cx="20"
        cy="20"
        fill="transparent"
        r="18.5"
        stroke="var(--theme-color-success-600)"
        stroke-dasharray="116.23892818282235"
        stroke-dashoffset="-2789.7342763877364"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="3"
        transform="rotate(-90 20 20)"
      />
      <text
        data-testid="CircularProgressBar__percentage"
        dy="0.3em"
        fill="var(--theme-color-success-600)"
        font-size="0.8rem"
        font-weight="600"
        text-anchor="middle"
        x="50%"
        y="50%"
      >
        <tspan>
          2500
          <tspan
            dy="-0.6em"
            font-size="0.4rem"
            font-weight="bold"
          >
            %
          </tspan>
        </tspan>
      </text>
    </svg>
  </div>
</div>
`;

exports[`PluginImage should render updating element in light mode 1`] = `
<div>
  <div
    class="PluginImage__PluginImageWrapper-sc-10kztrf-0 fBINnA bg-theme-primary-100 text-theme-primary-600 dark:bg-theme-secondary-800 dark:text-theme-secondary-700"
    data-testid="PluginImage"
  >
    <svg
      height="40"
      viewBox="0 0 40 40"
      width="40"
    >
      <circle
        cx="20"
        cy="20"
        fill="transparent"
        r="18.5"
        stroke="var(--theme-color-success-200)"
        stroke-width="3"
      />
      <circle
        cx="20"
        cy="20"
        fill="transparent"
        r="18.5"
        stroke="var(--theme-color-success-600)"
        stroke-dasharray="116.23892818282235"
        stroke-dashoffset="-2789.7342763877364"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="3"
        transform="rotate(-90 20 20)"
      />
      <text
        data-testid="CircularProgressBar__percentage"
        dy="0.3em"
        fill="var(--theme-color-success-600)"
        font-size="0.8rem"
        font-weight="600"
        text-anchor="middle"
        x="50%"
        y="50%"
      >
        <tspan>
          2500
          <tspan
            dy="-0.6em"
            font-size="0.4rem"
            font-weight="bold"
          >
            %
          </tspan>
        </tspan>
      </text>
    </svg>
  </div>
</div>
`;

exports[`PluginImage should render updating with label 1`] = `
<div>
  <div
    class="PluginImage__PluginImageWrapper-sc-10kztrf-0 fBINnA bg-theme-primary-100 text-theme-primary-600 dark:bg-theme-secondary-800 dark:text-theme-secondary-700"
    data-testid="PluginImage"
  >
    <svg
      height="40"
      viewBox="0 0 40 40"
      width="40"
    >
      <circle
        cx="20"
        cy="20"
        fill="transparent"
        r="18.5"
        stroke="var(--theme-color-success-800)"
        stroke-width="3"
      />
      <circle
        cx="20"
        cy="20"
        fill="transparent"
        r="18.5"
        stroke="var(--theme-color-success-600)"
        stroke-dasharray="116.23892818282235"
        stroke-dashoffset="-2789.7342763877364"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="3"
        transform="rotate(-90 20 20)"
      />
      <text
        data-testid="CircularProgressBar__percentage"
        dy="0.3em"
        fill="var(--theme-color-success-600)"
        font-size="0.8rem"
        font-weight="600"
        text-anchor="middle"
        x="50%"
        y="50%"
      >
        <tspan>
          2500
          <tspan
            dy="-0.6em"
            font-size="0.4rem"
            font-weight="bold"
          >
            %
          </tspan>
        </tspan>
      </text>
    </svg>
    <p
      class="text-sm font-semibold text-theme-success-600"
      data-testid="PluginImage__updating__label"
    >
      Updating...
    </p>
  </div>
</div>
`;

exports[`PluginImage should render with size '2xs' 1`] = `
<div>
  <div
    class="PluginImage__PluginImageWrapper-sc-10kztrf-0 gnJfJZ bg-theme-primary-100 text-theme-primary-600 dark:bg-theme-secondary-800 dark:text-theme-secondary-700"
    data-testid="PluginImage"
  >
    <svg>
      plugin-logo-placeholder.svg
    </svg>
  </div>
</div>
`;

exports[`PluginImage should render with size '3xs' 1`] = `
<div>
  <div
    class="PluginImage__PluginImageWrapper-sc-10kztrf-0 cKUSGF bg-theme-primary-100 text-theme-primary-600 dark:bg-theme-secondary-800 dark:text-theme-secondary-700"
    data-testid="PluginImage"
  >
    <svg>
      plugin-logo-placeholder.svg
    </svg>
  </div>
</div>
`;

exports[`PluginImage should render with size 'lg' 1`] = `
<div>
  <div
    class="PluginImage__PluginImageWrapper-sc-10kztrf-0 cvQNdU bg-theme-primary-100 text-theme-primary-600 dark:bg-theme-secondary-800 dark:text-theme-secondary-700"
    data-testid="PluginImage"
  >
    <svg>
      plugin-logo-placeholder.svg
    </svg>
  </div>
</div>
`;

exports[`PluginImage should render with size 'md' 1`] = `
<div>
  <div
    class="PluginImage__PluginImageWrapper-sc-10kztrf-0 cWrKPM bg-theme-primary-100 text-theme-primary-600 dark:bg-theme-secondary-800 dark:text-theme-secondary-700"
    data-testid="PluginImage"
  >
    <svg>
      plugin-logo-placeholder.svg
    </svg>
  </div>
</div>
`;

exports[`PluginImage should render with size 'sm' 1`] = `
<div>
  <div
    class="PluginImage__PluginImageWrapper-sc-10kztrf-0 TuOfP bg-theme-primary-100 text-theme-primary-600 dark:bg-theme-secondary-800 dark:text-theme-secondary-700"
    data-testid="PluginImage"
  >
    <svg>
      plugin-logo-placeholder.svg
    </svg>
  </div>
</div>
`;

exports[`PluginImage should render with size 'xs' 1`] = `
<div>
  <div
    class="PluginImage__PluginImageWrapper-sc-10kztrf-0 bpkNBH bg-theme-primary-100 text-theme-primary-600 dark:bg-theme-secondary-800 dark:text-theme-secondary-700"
    data-testid="PluginImage"
  >
    <svg>
      plugin-logo-placeholder.svg
    </svg>
  </div>
</div>
`;

exports[`PluginImage should show placeholder as fallback 1`] = `
<div>
  <div
    class="PluginImage__PluginImageWrapper-sc-10kztrf-0 cWrKPM bg-theme-primary-100 text-theme-primary-600 dark:bg-theme-secondary-800 dark:text-theme-secondary-700"
    data-testid="PluginImage"
  >
    <svg>
      plugin-logo-placeholder.svg
    </svg>
  </div>
</div>
`;
